src/job_cache/job_cache_impl_common.cpp

```diff
@@ -0,0 +1,251 @@
+  struct stat buf = {};
+  // There's a race here between the fstat and the copy_file_range
+  if (fstat(src_fd, &buf) < 0) {
+    log_fatal("fstat(src_fd = %d): %s", src_fd, strerror(errno));
+  }
+  if (copy_file_range(src_fd, nullptr, dst_fd, nullptr, buf.st_size, 0) < 0) {
+    log_fatal("copy_file_range(src_fd = %d, NULL, dst_fd = %d, size = %d, 0): %s", src_fd, dst_fd,
+              buf.st_size, strerror(errno));
+  }
+}
+
+// For older linux distros use Linux's sendfile
+#else
+
+#include <sys/sendfile.h>
+// This function just uses `sendfile` to make
+// an efficent copy. It is however not atomic because
+// we have to `fstat` before we call `sendfile`.
+// This means that if an external party decides to mutate
+// the file (espeically changing its size) then this function
+// will not work as intended. External parties *are* allowed to
+// unlink this file but they're not allowed to modify it in any
+// other way or else this function will race with that external
+// modification.
+
+static void copy(int src_fd, int dst_fd) {
+  struct stat buf = {};
+  // There's a race here between the fstat and the copy_file_range
+  if (fstat(src_fd, &buf) < 0) {
+    log_fatal("fstat(src_fd = %d): %s", src_fd, strerror(errno));
+  }
+  off_t idx = 0;
+  size_t size = buf.st_size;
+  do {
+    intptr_t written = sendfile(dst_fd, src_fd, &idx, size);
+    if (written < 0) {
+      log_fatal("sendfile(src_fd = %d, NULL, dst_fd = %d, size = %d, 0): %s", src_fd, dst_fd,
+                buf.st_size, strerror(errno));
+    }
+    idx = written;
+    size -= written;
+  } while (size != 0);
+}
+#endif
+
+#ifdef FICLONE
+
+void copy_or_reflink(const char *src, const char *dst, mode_t mode, int extra_flags) {
+  auto src_fd = wcl::unique_fd::open(src, O_RDONLY);
+  if (!src_fd) {
+    log_fatal("open(%s): %s", src, strerror(src_fd.error()));
+  }
+  auto dst_fd = wcl::unique_fd::open(dst, O_WRONLY | O_CREAT | extra_flags, mode);
+  if (!dst_fd) {
+    log_fatal("open(%s): %s", dst, strerror(dst_fd.error()));
+  }
+
+  if (ioctl(dst_fd->get(), FICLONE, src_fd->get()) < 0) {
+    if (errno != EINVAL && errno != EOPNOTSUPP && errno != EXDEV) {
+      log_fatal("ioctl(%s, FICLONE, %d): %s", dst, src, strerror(errno));
+    }
+    copy(src_fd->get(), dst_fd->get());
+  }
+}
+
+#else
+
+void copy_or_reflink(const char *src, const char *dst, mode_t mode, int extra_flags) {
+  auto src_fd = wcl::unique_fd::open(src, O_RDONLY);
+  if (!src_fd) {
+    log_fatal("open(%s): %s", src, strerror(src_fd.error()));
+  }
+  auto dst_fd = wcl::unique_fd::open(dst, O_WRONLY | O_CREAT | extra_flags, mode);
+  if (!dst_fd) {
+    log_fatal("open(%s): %s", dst, strerror(dst_fd.error()));
+  }
+
+  copy(src_fd->get(), dst_fd->get());
+}
+
+#endif
+
+void remove_backing_files(const std::string &dir, int64_t job_id) {
+  uint8_t group_id = job_id & 0xFF;
+  std::string job_dir = wcl::join_paths(dir, wcl::to_hex(&group_id), std::to_string(job_id));
```

Comment:
This looks like duplicated code from somewhere else right? Like is this code also used somewhere for determining where to write a file?  If so can we make a "get job_dir from job_id" function so that we don't have to keep that code aligned everywhere we do it?